x-pack/plugins/security_solution/public/timelines/components/timeline/unified_components/data_table/custom_timeline_data_grid_body.tsx

```diff
@@ -101,11 +113,37 @@ const CustomDataGridSingleRow = memo(function CustomDataGridSingleRow(
 ) {
   const { rowIndex, rowData, enabledRowRenderers, visibleColumns, Cell } = props;
 
+  const [intersectionEntry, setIntersectionEntry] = useState<IntersectionObserverEntry>({
+    isIntersecting: false,
+    intersectionRatio: 0,
+  } as IntersectionObserverEntry);
+
+  const intersectionRef = useRef<HTMLDivElement | null>(null);
+
+  const observer = useRef<IntersectionObserver | null>(null);
+
   const { canShowRowRenderer } = useStatefulRowRenderer({
     data: rowData.ecs,
     rowRenderers: enabledRowRenderers,
   });
 
+  useEffect(() => {
+    if (intersectionRef.current && !observer.current) {
```

Comment:
I don't want this to be over-engineered, but once a row's been loaded, we shouldn't have to load it again which is what's happening now from what I see. Not sure if we can keep a map of loaded rows that only resets when it should (i.e. query change, column change, per page count, etc...)

Reply:
So that was my first go to strategy but it leads to slew of new problems.. Once 100 rows has been already loaded, the UI becomes extremely sluggish for any operations such as `minimizing/maximizing timeline` or switching tabs.

Additionaly, if rows are loaded in EQL and ESQL tab, then there are 200 rows rendered at the same time and performance degrades even more.


Showing only the rows that are visible to user is  something new but exactly how virtualized table works, which was what I was trying to emulate. Checkout how discover works in below video 

https://github.com/elastic/kibana/assets/7485038/62b6b664-d55a-47cb-91d4-d5f2d291d525

Reply:
>Not sure if we can keep a map of loaded rows that only resets when it should (i.e. query change, column change, per page count, etc...)

Problem is not the data but it is rendering. Data is already available.